internal/k8s/controllers/service_controller.go

```diff
@@ -0,0 +1,145 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+
+	"go.universe.tf/metallb/internal/k8s/epslices"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ServiceReconciler struct {
+	client.Client
+	Logger      log.Logger
+	Scheme      *runtime.Scheme
+	Namespace   string
+	Handler     func(log.Logger, string, *v1.Service, epslices.EpsOrSlices) SyncState
+	Endpoints   NeedEndPoints
+	ForceReload func()
+}
+
+//+kubebuilder:rbac:groups="",resources=service,verbs=get;list;watch;
+//+kubebuilder:rbac:groups="",resources=service/status,verbs=update;
+//+kubebuilder:rbac:groups="",resources=endpoints,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=discovery.k8s.io,resources=endpointslices,verbs=get;list;watch;
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;patch;
+
+func (r *ServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "ServiceReconciler", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "ServiceReconciler", "end reconcile", req.NamespacedName.String())
+
+	var service *v1.Service
+
+	service, err := r.serviceFor(ctx, req.NamespacedName)
+	if err != nil {
+		level.Error(r.Logger).Log("controller", "ServiceReconciler", "error", "failed to get service", "service", req.NamespacedName, "error", err)
+		return ctrl.Result{}, err
+	}
+
+	epSlices, err := epsOrSlicesForServices(ctx, r, req.NamespacedName, r.Endpoints)
+	if err != nil {
+		level.Error(r.Logger).Log("controller", "ServiceReconciler", "error", "failed to get endpoints", "service", req.NamespacedName, "error", err)
+		return ctrl.Result{}, err
+	}
+	if service != nil {
+		level.Debug(r.Logger).Log("controller", "ServiceReconciler", "processing service", spew.Sdump(service))
+	} else {
```

Comment:
when do we get here?
if the service is nil it means there was an error we found when getting the service and returned.

Reply:
No, it means that we are handling a delete event and serviceFor above returned nil (which still is a valid event).
The next line also says that

Reply:
oh ok i now see it in the serviceFor func